File: 809/809_version2.java
class Solution {
    private String target;
    public int expressiveWords(String S, String[] words) {
        int size = words.length;
        if(size==0 || S.length()==0)
            return 0;
        target = S;
        int result = 0;
        for(String word:words)
            if(compareWord(word)) result++;
        return result;
    }
    private boolean compareWord(String source){
        int sLen = source.length(),tLen = target.length();
        int sstart=0,send=0;
        int tstart=0,tend=0;
        for(send=0,tend=0;sstart<sLen&&tstart<tLen;sstart=send,tstart=tend){
            if(source.charAt(sstart)!=target.charAt(tstart)) return false;
            while(send<sLen&&source.charAt(sstart)==source.charAt(send)) send++;
            while(tend<tLen&&target.charAt(tstart)==target.charAt(tend)) tend++;
            int sDis = send-sstart,tDis = tend-tstart;
            if(sDis!=tDis && tDis < Math.max(3,sDis)) return false;
        }
        return sstart==sLen && tstart==tLen;
    }
}

File: 809/809_version2.py
class Solution:
    def expressiveWords(self, S: str, words: list) -> int:
        self.target = S
        if not self.target or not words:
            return 0
        return sum(self.check(word) for word in words)
    def check(self, source:str):
        s_start,s_end,t_start,t_end = 0,0,0,0
        s_len,t_len = len(source),len(self.target)
        while s_end < s_len and t_end < t_len:
            if self.target[t_start]!=source[s_start]:
                return False
            while s_end < s_len and source[s_end]==source[s_start]:
                s_end+=1
            while t_end < t_len and self.target[t_end]==self.target[t_start]:
                t_end+=1
            s_dis,t_dis = s_end-s_start,t_end-t_start
            if s_dis!=t_dis and t_dis < max([3,s_dis]):return False
            s_start,t_start = s_end,t_end
        return s_start==s_len and t_start==t_len
